locale.c

```diff
@@ -251,225 +251,282 @@ static const char C_thousands_sep[] = "";
-};
-
-/* The top-most real element is LC_ALL */
-
-STATIC const char * const category_names[] = {
-
-#    ifdef USE_LOCALE_CTYPE
-                                 "LC_CTYPE",
-#    endif
-#  ifdef USE_LOCALE_NUMERIC
-                                 "LC_NUMERIC",
-#  endif
-#    ifdef USE_LOCALE_COLLATE
-                                 "LC_COLLATE",
-#    endif
-#    ifdef USE_LOCALE_TIME
-                                 "LC_TIME",
-#    endif
-#    ifdef USE_LOCALE_MESSAGES
-                                 "LC_MESSAGES",
-#    endif
-#    ifdef USE_LOCALE_MONETARY
-                                 "LC_MONETARY",
-#    endif
-#    ifdef USE_LOCALE_ADDRESS
-                                 "LC_ADDRESS",
-#    endif
-#    ifdef USE_LOCALE_IDENTIFICATION
-                                 "LC_IDENTIFICATION",
-#    endif
-#    ifdef USE_LOCALE_MEASUREMENT
-                                 "LC_MEASUREMENT",
-#    endif
-#    ifdef USE_LOCALE_PAPER
-                                 "LC_PAPER",
-#    endif
-#    ifdef USE_LOCALE_TELEPHONE
-                                 "LC_TELEPHONE",
-#    endif
-#    ifdef USE_LOCALE_NAME
-                                 "LC_NAME",
-#    endif
-#    ifdef USE_LOCALE_SYNTAX
-                                 "LC_SYNTAX",
-#    endif
-#    ifdef USE_LOCALE_TOD
-                                 "LC_TOD",
-#    endif
-#    ifdef LC_ALL
-                                 "LC_ALL",
-#    else
-                                 "If you see this, it is a bug in"
-                                 " perl; please report it via perlbug",
-#    endif
-                                 "Locale category unknown to Perl; if"  \
-                                 " you see this, it is a bug in perl;"  \
-                                 " please report it via perlbug"
-};
+/* Defines for the parallel arrays and other items below.  There is a single
+ * set for every possible locale category on any system.  If you add categories
+ * be AWARE that Configure will need to change to include the new one.
+ *
+ * If the category is present on the system, each symbol expands to the
+ * appropriate value.  If not present, the symbol expands to nothing.  This
+ * allows each array to list all symbols, and only the ones relevant to the
+ * system actually have content.  This way of formatting these brings all the
+ * varying components adjacent, so a glance tells if something is awry. */
 
-/* A few categories require additional setup when they are changed.  This table
- * points to the functions that do that setup */
-STATIC void (*update_functions[]) (pTHX_ const char *, bool force) = {
 #  ifdef USE_LOCALE_CTYPE
-                                S_new_ctype,
+#    define LC_CTYPE_ENTRY_                      LC_CTYPE,
+#    define LC_CTYPE_MASK_                       LC_CTYPE_MASK,
+#    define LC_CTYPE_STRING_                    "LC_CTYPE",
+#      define LC_CTYPE_UPDATE_          S_new_ctype,
+#  else
```

Comment:
Would this series of commits be more maintainable if it worked something like `intrpvar.h`?

It would significantly reduce the repetition: how many times is `LC_CTYPE` repeated here, including as substrings of macro names?

For example locale_table.h could be:
```
#ifdef USE_LOCALE_CTYPE
  LOCALE_TABLE_ENTRY(LC_CTYPE, LC_CTYPE_MASK, "LC_CTYPE", S_new_ctype)
#endif
#ifdef USE_LOCALE_NUMERIC
  LOCALE_TABLE_ENTRY(LC_NUMERIC, LC_NUMERIC_MASK, "LC_NUMERIC", S_new_numeric)
#endif
... and more
```
Then in locale.c:
```
#undef LOCALE_TABLE_ENTRY
#define LOCALE_TABLE_ENTRY(index, mask, name, cb) index,
STATIC const int categories[] = {
#include "locale_table.h"
  FAKE_LCALL+1
}
#undef LOCALE_TABLE_ENTRY
#defined LOCALE_TABLE_ENTRY(index, mask, name, cb) name,
STATIC const char * const category_names[] = {
#include "locale_table.h"
   "Locale category unknown to Perl; ..."
}
... etc
```
though the mask and name could be synthesized from the index parameter.

Reply:
This sounds like a good idea.  I'll work on it.